File: Daily/2616.minimize-the-maximum-difference-of-pairs.js
/*
 * @lc app=leetcode id=2616 lang=javascript
 *
 * [2616] Minimize the Maximum Difference of Pairs
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @param {number} p
 * @return {number}
 */
var minimizeMax = function (nums, p) {
    nums.sort((a, b) => a - b);

    /**
     * @param {number} diff
     * @returns {number}
     */
    const getPairNumBelowDiff = (diff) => {
        let count = 0;
        for (let i = 0; i < nums.length - 1; i++) {
            if (nums[i + 1] - nums[i] <= diff) {
                count++;
                i++;
            }
        }
        return count;
    };

    let leftDiff = 0;
    let rightDiff = nums[nums.length - 1] - nums[0];

    while (leftDiff <= rightDiff) {
        const midDiff = Math.floor((rightDiff - leftDiff) / 2) + leftDiff;

        const pairNumBelowDiff = getPairNumBelowDiff(midDiff);
        if (pairNumBelowDiff < p) {
            leftDiff = midDiff + 1;
        } else if (pairNumBelowDiff >= p) {
            if (midDiff === 0 || getPairNumBelowDiff(midDiff - 1) < p) {
                return midDiff;
            } else {
                rightDiff = midDiff - 1;
            }
        }
    }
};
// @lc code=end

minimizeMax([10, 1, 2, 7, 1, 3], 3);

File: src/2025/2616.minimize-the-maximum-difference-of-pairs.cpp
/*
 * @lc app=leetcode id=2616 lang=cpp
 *
 * [2616] Minimize the Maximum Difference of Pairs
 */

#include <algorithm>
#include <vector>

// @lc code=start
class Solution {
 public:
  int minimizeMax(std::vector<int>& nums, const int p) {
    if (p == 0) {
      return 0;
    }
    std::ranges::sort(nums);

    int left = 0;
    int right = nums.back() - nums.front();

    while (left < right) {
      const int mid = left + (right - left) / 2;
      const int pair_count = GetPairCountWithMaximumDiff(nums, mid);
      if (pair_count < p) {
        left = mid + 1;
      } else if (pair_count >= p) {
        right = mid;
      }
    }

    return left;
  }

 private:
  static int GetPairCountWithMaximumDiff(const std::vector<int>& nums,
                                         const int maximum_diff) {
    int current_pair_number = 0;
    int index = 0;

    while (index < nums.size() - 1) {
      if (nums[index + 1] - nums[index] > maximum_diff) {
        index++;
      } else {
        current_pair_number++;
        index += 2;
      }
    }

    return current_pair_number;
  }
};
// @lc code=end
